Browser_IDE/IDBStorage.js

```diff
@@ -0,0 +1,265 @@
+            await RW.closeDB();
+        }
+    }
+}
+
+class __IDBStorageRW{
+    constructor(IDBS) {
+        this.owner = IDBS;
+        this.db = null;
+        this.doInitialization = false;
+        this.performedWrite = false;
+    }
+
+    openDB(){
+        let IDBS = this;
+        return new Promise((resolve, reject)=>{
+            if (IDBS.db != null)
+                reject();
+
+            let req = indexedDB.open("SplashKitOnline", 1);
+
+            req.onupgradeneeded = async (e) => {
+                IDBS.db = req.result;
+                IDBS.db.createObjectStore("app", {keyPath: "category"});
+                
+                let userProjectsStore = IDBS.db.createObjectStore("userProjects", {keyPath: "id"});
+                userProjectsStore.createIndex("name", "name", {unique: true});
+
+                if (e.oldVersion == 0)
+                    IDBS.doInitialization = true;
+            };
+
+            req.onsuccess = async (e) => {
+                IDBS.db = req.result;
+                if (IDBS.doInitialization){
+                    await IDBS.updateLastWriteTime();
+                }
+                IDBS.doInitialization = false;
+                resolve();
+            };
+
+            req.onerror = async (e) => {
+                IDBS.owner.dispatchEvent(new Event("connectionFailed"));
+                reject();
+            };
+        });
+    }
+
+    async closeDB(){
+        if (this.performedWrite)
+            this.updateLastWriteTime();
+        if (this.db != null)
+            this.db.close();
+        this.db = null;
+    }
+
+    doTransaction(store, state, func)
+    {
+        let IDBS = this;
+        return new Promise(async (resolve, reject) => {
+            let transaction = IDBS.db.transaction(store, state);
+            let files = transaction.objectStore(store);
+            let result = undefined;
+
+            try {
+                result = await func(transaction, files);
+            } catch(err){
+                reject(err);
+                return;
+            }
+
+            transaction.onerror = function(){
+                console.log("error");
+                transaction.abort(); 
+                reject(transaction.error);
+            };
+            
+            transaction.oncomplete = function(){
+                resolve(result);
+            };
+        });
+    }
+
+    request(transaction, func)
+    {
+        return new Promise(async (resolve, reject) => {
+            let result = undefined;
+            
+            try {
+                result = await func();
+            } catch(err){
+                reject(err);
+                return;
+            }
+
+            result.onerror = function(){
+                console.log("error");
+                transaction.abort(); 
+                reject(result.error);
+            };
+
+            result.onsuccess = function(){
+                resolve(result.result);
+            };
+        });
+    }

```

Comment:
This is just a question/thought, but I wonder if there's a way to generalize some of this. Maybe it's a bit early (rule of three and all), but I worry we'll end up duplicating this structure (storage, the private _storageRW, doTransation+request) each time we add a new database 'type'. No need to change this, just wanted to bring it up in case you had any thoughts here.

Reply:
I began an attempt to abstract it but decided the refactoring would be outside the scope of the task by somewhat too much. It's definitely a good idea, and would make it easier to reason about and modify the business logic of these objects, which is currently strewn across several callbacks.

Reply:
Yeah that's fair enough. I'm happy with it as-is for now and might just add another task for abstracting it later on 😄